domain-check.bash

```diff
@@ -0,0 +1,148 @@
+# enough capabilities inside of the urlgetter pseudo-test and (2)
+# prototype the next generation Web Connectivity.
+#
+# Very alpha. Very basic. Needs much more love.
+#
+# Inspired by great chats and great code. Main source of ispiration is
+# currently https://github.com/Jigsaw-Code/net-analysis/blob/master/netanalysis/blocktest/measure.sh
+
+function fatal() {
+  echo "FATAL: $@" 1>&2
+  exit 1
+}
+
+if [ $# -ne 1 ]; then
+  fatal "usage: $0 domain"
+fi
+domain=$1
+
+function require() {
+  if ! [ -x "$(command -v $1)" ]; then
+    fatal "$1 not installed, please run: $2"
+  fi
+}
+
+require jq "sudo apt install jq"
+require uuidgen "sudo apt install uuid-runtime"
+
+uuid=$(uuidgen)
+
+function log() {
+  echo "$@" 1>&2
+}
+
+function run() {
+  log "$@"
+  "$@"
+}
+
+function urlgetter() {
+  #run ./miniooni -A session=$uuid $@ urlgetter
+  run ./miniooni -n --no-bouncer -A session=$uuid "$@" urlgetter
+}
+
+function getfailure() {
+  tail -n1 report.jsonl|jq -r .test_keys.failure
+}
+
+function output() {
+  echo "$@"
+}
+
+function getipv4first() {
+  tail -n1 report.jsonl|jq -r ".test_keys.queries|.[]|select(.hostname==\"$1\")|select(.query_type==\"A\")|.answers|.[0].ipv4"
+}
+
+function getipv4list() {
+  tail -n1 report.jsonl|jq -r ".test_keys.queries|.[]|select(.hostname==\"$1\")|select(.query_type==\"A\")|.answers|.[].ipv4"|sort
+}
+
+function getbody() {
+  # Implementation note: requests stored in LIFO order
+  tail -n1 report.jsonl|jq -r ".test_keys.requests|.[0]|.response.body"
+}
+
+testdomain=${MINIOONI_TEST_DOMAIN:-example.org}
+
+log "* getting IP address of $testdomain"
+urlgetter -ODNSCache="8.8.4.4 dns.google" -OResolverURL=doh://google -i dnslookup://$testdomain
+if [ "$(getfailure)" != "null" ]; then
+  fatal "cannot determine IP address of $testdomain"
+fi
+testip=$(getipv4first $testdomain)
+if [ -z "$testip" ]; then
+  fatal "no available IPv4 for $testdomain"
+fi
+output "MINIOONI_TEST_DOMAIN=$testdomain"
+output "MINIOONI_TEST_IP=$testip"
+
+log "* checking for sni-triggered blocking"
+urlgetter -OTLSServerName=$domain -i tlshandshake://$testip:443
+if [ "$(getfailure)" != "ssl_invalid_hostname" ]; then
+  output "MINIOONI_SNI_BLOCKING=1"
+fi
+
+log "* checking for host-header-triggered censorship"
+urlgetter -OHTTPHost=$domain -ONoFollowRedirects=true -i http://$testip
+if [ "$(getfailure)" != "null" ]; then
+  output "MINIOONI_HOST_BLOCKING=1"
+fi
+
+log "* checking for DNS injection"
+urlgetter -OResolverURL=udp://$testip:53 -i dnslookup://$domain
+if [ "$(getfailure)" = "null" ]; then
+  output "MINIOONI_DNS_INJECTION=1"
+fi
+
+log "* resolving $domain using the system resolver"
+# Implementation note: this saves the IPs _and_ records the error.
+# XXX: maybe make these two checks separate checks for now?
+urlgetter -OResolverURL=system:/// -ORejectDNSBogons=true -i dnslookup://$domain
+if [ "$(getfailure)" = "dns_bogon_error" ]; then
+  output "MINIOONI_DNS_BOGONS=1"
```

Comment:
This doesn't really explain the censorship mechanism.
Was the bogon injected or did the resolver lie?
That's what matters the most. The bogon detection is just a way to validate the IP address, but the result should be DNS lie or DNS injection.

Reply:
I agree. I think this script was an initial experiment to understand how an higher level logic for running measurements (i.e. an improved web connectivity) could work. For this reason, I didn't bother with implementing an algorithm for providing output to users of the script. The goal is to understand a good algorithm and then implement it in Go.